File: src/python_pkg/python_pkg/imu_transform.py
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Imu
from geometry_msgs.msg import TransformStamped
from tf2_ros import StaticTransformBroadcaster
import numpy as np
import yaml
import os

class ImuTransform(Node):
    def __init__(self):
        super().__init__('imu_transform_node')
        self.get_logger().info("IMU 转换节点已启动")
        
        # 声明参数
        self.declare_parameters(
            namespace='',
            parameters=[
                ('imu_topic', '/imu_raw'),
                ('imu_transformed_topic', '/imu_lidar_frame'),
                ('imu_frame', 'imu_link'),
                ('lidar_frame', 'lidar_link'),
                ('Calibration_file', 'config/imu_transform.yaml'),
                ('pub_tf', True),
                ('use_transform', True),
                ('use_grivaty2m', False)
            ]
        )
        
        # 获取参数
        self.imu_topic = self.get_parameter('imu_topic').value
        self.imu_transformed_topic = self.get_parameter('imu_transformed_topic').value
        self.imu_frame = self.get_parameter('imu_frame').value
        self.lidar_frame = self.get_parameter('lidar_frame').value
        calib_file = self.get_parameter('Calibration_file').value
        
        # 订阅和发布
        self.imu_sub = self.create_subscription(
            Imu, 
            self.imu_topic, 
            self.imu_callback, 
            10
        )
        self.imu_pub = self.create_publisher(Imu, self.imu_transformed_topic, 10)
        
        # 初始化变换矩阵
        self.imu_to_lidar = self.parse_calibration_file(calib_file)
        self._gravity = 9.81
        
        # 初始化 TF 广播
        if self.get_parameter('pub_tf').value:
            self.static_broadcaster = StaticTransformBroadcaster(self)
            self.publish_static_tf()

    def parse_calibration_file(self, file_path):
        """支持解析 YAML/TXT 文件"""
        if not os.path.exists(file_path):
            self.get_logger().error(f"校准文件不存在: {file_path}")
            raise FileNotFoundError(f"Calibration file {file_path} not found")

        matrix = np.eye(4)
        try:
            if file_path.endswith(('.yaml', '.yml')):
                with open(file_path) as f:
                    data = yaml.safe_load(f)
                    matrix = np.array(data['transformation_matrix'])
            else:  # 处理 TXT 文件
                with open(file_path) as f:
                    matrix_started = False
                    row = 0
                    for line in f:
                        if "Homogeneous Transformation Matrix from LiDAR to IMU" in line:
                            matrix_started = True
                            continue
                        if matrix_started and row < 4:
                            vals = list(map(float, line.strip().split()))
                            matrix[row] = vals[:4]
                            row += 1
        except Exception as e:
            self.get_logger().error(f"解析校准文件失败: {str(e)}")
            raise

        return np.linalg.inv(matrix)

    def imu_callback(self, msg):
        transformed_msg = Imu()
        transformed_msg.header = msg.header
        
        # 处理重力单位转换
        if self.get_parameter('use_grivaty2m').value:
            transformed_msg.linear_acceleration.x = msg.linear_acceleration.x * self._gravity
            transformed_msg.linear_acceleration.y = msg.linear_acceleration.y * self._gravity
            transformed_msg.linear_acceleration.z = msg.linear_acceleration.z * self._gravity
        else:
            transformed_msg.linear_acceleration = msg.linear_acceleration
            
        # 应用坐标变换
        if self.get_parameter('use_transform').value:
            # 加速度变换
            acc = np.array([
                msg.linear_acceleration.x,
                msg.linear_acceleration.y,
                msg.linear_acceleration.z,
                0
            ])
            acc_transformed = self.imu_to_lidar @ acc
            transformed_msg.linear_acceleration.x = acc_transformed[0]
            transformed_msg.linear_acceleration.y = acc_transformed[1]
            transformed_msg.linear_acceleration.z = acc_transformed[2]
            
            # 角速度变换
            gyro = np.array([
                msg.angular_velocity.x,
                msg.angular_velocity.y,
                msg.angular_velocity.z,
                0
            ])
            gyro_transformed = self.imu_to_lidar @ gyro
            transformed_msg.angular_velocity.x = gyro_transformed[0]
            transformed_msg.angular_velocity.y = gyro_transformed[1]
            transformed_msg.angular_velocity.z = gyro_transformed[2]
        
        # 更新坐标系
        transformed_msg.header.frame_id = self.lidar_frame
        self.imu_pub.publish(transformed_msg)

    def publish_static_tf(self):
        transform = TransformStamped()
        transform.header.stamp = self.get_clock().now().to_msg()
        transform.header.frame_id = self.imu_frame
        transform.child_frame_id = self.lidar_frame
        
        # 设置平移
        transform.transform.translation.x = self.imu_to_lidar[0, 3]
        transform.transform.translation.y = self.imu_to_lidar[1, 3]
        transform.transform.translation.z = self.imu_to_lidar[2, 3]
        
        # 计算四元数
        rotation_matrix = self.imu_to_lidar[:3, :3]
        q = self.matrix_to_quaternion(rotation_matrix)
        transform.transform.rotation.x = q[0]
        transform.transform.rotation.y = q[1]
        transform.transform.rotation.z = q[2]
        transform.transform.rotation.w = q[3]
        
        self.static_broadcaster.sendTransform(transform)

    @staticmethod
    def matrix_to_quaternion(matrix):
        """将旋转矩阵转换为四元数"""
        m = matrix.T  # ROS 使用主动变换矩阵
        trace = np.trace(m)
        if trace > 0:
            S = np.sqrt(trace + 1.0) * 2
            qw = 0.25 * S
            qx = (m[2, 1] - m[1, 2]) / S
            qy = (m[0, 2] - m[2, 0]) / S
            qz = (m[1, 0] - m[0, 1]) / S
        elif (m[0, 0] > m[1, 1]) and (m[0, 0] > m[2, 2]):
            S = np.sqrt(1.0 + m[0, 0] - m[1, 1] - m[2, 2]) * 2
            qw = (m[2, 1] - m[1, 2]) / S
            qx = 0.25 * S
            qy = (m[0, 1] + m[1, 0]) / S
            qz = (m[0, 2] + m[2, 0]) / S
        elif m[1, 1] > m[2, 2]:
            S = np.sqrt(1.0 + m[1, 1] - m[0, 0] - m[2, 2]) * 2
            qw = (m[0, 2] - m[2, 0]) / S
            qx = (m[0, 1] + m[1, 0]) / S
            qy = 0.25 * S
            qz = (m[1, 2] + m[2, 1]) / S
        else:
            S = np.sqrt(1.0 + m[2, 2] - m[0, 0] - m[1, 1]) * 2
            qw = (m[1, 0] - m[0, 1]) / S
            qx = (m[0, 2] + m[2, 0]) / S
            qy = (m[1, 2] + m[2, 1]) / S
            qz = 0.25 * S
        return np.array([qx, qy, qz, qw])

def main(args=None):
    rclpy.init(args=args)
    node = ImuTransform()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/my_perception/src/imu_transform.cpp
#include <Eigen/Dense>
#include <fstream>
#include <geometry_msgs/msg/transform_stamped.hpp>
#include <rclcpp/rclcpp.hpp>
#include <rclcpp_components/register_node_macro.hpp>
#include <sensor_msgs/msg/imu.hpp>
#include <sstream>
#include <tf2_ros/static_transform_broadcaster.h>
#include <yaml-cpp/yaml.h>

class ImuTransform : public rclcpp::Node
{
public:
    ImuTransform(const rclcpp::NodeOptions &options)
        : Node("imu_transform_node", options)
    {
        RCLCPP_INFO(this->get_logger(), "IMU 转换节点已启动");
        // 读取 YAML 配置
        std::string yaml_file = "config/imu_transform.yaml";
        this->declare_parameter("imu_topic", "/imu_raw");
        this->declare_parameter("imu_transformed_topic", "/imu_lidar_frame");
        this->declare_parameter("imu_frame", "imu_link");
        this->declare_parameter("lidar_frame", "lidar_link");
        this->declare_parameter("Calibration_file", yaml_file);
        this->declare_parameter("pub_tf", true);
        this->declare_parameter("use_transform", true);
        this->declare_parameter("use_grivaty2m", false);
        imu_topic_ = this->get_parameter("imu_topic").as_string();
        imu_transformed_topic_ = this->get_parameter("imu_transformed_topic").as_string();
        imu_frame_ = this->get_parameter("imu_frame").as_string();
        lidar_frame_ = this->get_parameter("lidar_frame").as_string();
        // 订阅 IMU 话题
        imu_sub_ = this->create_subscription<sensor_msgs::msg::Imu>(
            imu_topic_, 10, [this](const sensor_msgs::msg::Imu::SharedPtr msg)
            { imuCallback(msg); });

        // 发布转换后的 IMU 话题
        imu_pub_ = this->create_publisher<sensor_msgs::msg::Imu>(imu_transformed_topic_, 10);
        parseYaml(get_parameter("Calibration_file").as_string());
        // 发布 TF 变换
        if (this->get_parameter("pub_tf").as_bool())
        {
            static_broadcaster_ = std::make_shared<tf2_ros::StaticTransformBroadcaster>(this);
            publishStaticTF();
        }
    }

private:
    rclcpp::Subscription<sensor_msgs::msg::Imu>::SharedPtr imu_sub_;
    rclcpp::Publisher<sensor_msgs::msg::Imu>::SharedPtr imu_pub_;
    std::shared_ptr<tf2_ros::StaticTransformBroadcaster> static_broadcaster_;
    const double _gravity = 9.81;
    std::string imu_topic_, imu_transformed_topic_, imu_frame_, lidar_frame_;
    Eigen::Matrix4d imu_to_lidar_;

    void parseYaml(const std::string &file_path)
    {
        std::ifstream file(file_path);
        if (!file.is_open())
        {
            RCLCPP_ERROR(this->get_logger(), "无法打开 YAML 文件: %s", file_path.c_str());
            rclcpp::shutdown();
        }

        Eigen::Matrix4d lidar_to_imu;
        std::string line;
        bool matrix_started = false;
        int row = 0;

        while (std::getline(file, line))
        {
            // 识别矩阵部分
            if (line.find("Homogeneous Transformation Matrix from LiDAR to IMU") != std::string::npos)
            {
                matrix_started = true;
                row = 0;
                continue;
            }

            if (matrix_started)
            {
                std::istringstream iss(line);
                double val;
                int col = 0;

                while (iss >> val)
                {
                    lidar_to_imu(row, col) = val;
                    col++;
                }

                row++;
                if (row >= 4)
                    break; // 只解析 4 行
            }
        }

        file.close();

        // 计算 IMU 到 LiDAR 变换矩阵（逆矩阵）
        imu_to_lidar_ = lidar_to_imu.inverse();
        RCLCPP_INFO(this->get_logger(), "IMU to LiDAR 变换矩阵计算完成");
    }

    void imuCallback(const sensor_msgs::msg::Imu::SharedPtr msg)
    {
        sensor_msgs::msg::Imu transformed_msg = *msg;
        if (get_parameter("use_grivaty2m").as_bool())
        {
            transformed_msg.linear_acceleration.z *= _gravity;
            transformed_msg.linear_acceleration.x *= _gravity;
            transformed_msg.linear_acceleration.y *= _gravity;
        }
        if (get_parameter("use_transform").as_bool())
        {
            Eigen::Vector3d acc(msg->linear_acceleration.x, msg->linear_acceleration.y, msg->linear_acceleration.z);
            Eigen::Vector3d acc_transformed = imu_to_lidar_.block<3, 3>(0, 0) * acc;
            transformed_msg.linear_acceleration.x = acc_transformed.x();
            transformed_msg.linear_acceleration.y = acc_transformed.y();
            transformed_msg.linear_acceleration.z = acc_transformed.z();

            // 旋转 IMU 角速度数据
            Eigen::Vector3d gyro(msg->angular_velocity.x, msg->angular_velocity.y, msg->angular_velocity.z);
            Eigen::Vector3d gyro_transformed = imu_to_lidar_.block<3, 3>(0, 0) * gyro;
            transformed_msg.angular_velocity.x = gyro_transformed.x();
            transformed_msg.angular_velocity.y = gyro_transformed.y();
            transformed_msg.angular_velocity.z = gyro_transformed.z();
        }
        // 旋转 IMU 加速度数据

        // 修改坐标系
        transformed_msg.header.frame_id = lidar_frame_;
        transformed_msg.header.stamp = msg->header.stamp;

        imu_pub_->publish(transformed_msg);
    }

    void publishStaticTF()
    {
        geometry_msgs::msg::TransformStamped transformStamped;

        transformStamped.header.stamp = this->now();
        transformStamped.header.frame_id = imu_frame_;
        transformStamped.child_frame_id = lidar_frame_;

        transformStamped.transform.translation.x = imu_to_lidar_(0, 3);
        transformStamped.transform.translation.y = imu_to_lidar_(1, 3);
        transformStamped.transform.translation.z = imu_to_lidar_(2, 3);

        Eigen::Matrix3d rotation_matrix = imu_to_lidar_.block<3, 3>(0, 0);
        Eigen::Quaterniond q(rotation_matrix);

        transformStamped.transform.rotation.x = q.x();
        transformStamped.transform.rotation.y = q.y();
        transformStamped.transform.rotation.z = q.z();
        transformStamped.transform.rotation.w = q.w();

        static_broadcaster_->sendTransform(transformStamped);
    }
};

// int main(int argc, char **argv)
// {
//     rclcpp::init(argc, argv);
//     rclcpp::spin(std::make_shared<ImuTransformNode>());
//     rclcpp::shutdown();
//     return 0;
// }
RCLCPP_COMPONENTS_REGISTER_NODE(ImuTransform)